cusub_perception/localizer/src/start_gate_watershed.cpp

```diff
@@ -7,17 +7,206 @@
+            if(l[3] < top_pt.y)
+                top_pt = Point2f(l[2],l[3]);
+        }
+        points.push_back(bottom_pt);
+        points.push_back(top_pt);
+        return true;
+    }
+
     bool StartGateWatershed::generatePose(
         sensor_msgs::Image& image, 
-        std::vector<darknet_ros_msgs::BoundingBox>& bbs,
+        vector<darknet_ros_msgs::BoundingBox>& bbs,
         geometry_msgs::Pose& pose,
-        std::string& class_name
+        string& class_name
     ){
+        if(bbs.size() != 2) {return false;} // to be adjusted for 3 leg case
+        sortBoxes(bbs);
         ROS_INFO("Localizing the gate!");
+        // Double check that its RGB8 not BGR8
+        cv_bridge::CvImagePtr cv_ptr = cv_bridge::toCvCopy(image, sensor_msgs::image_encodings::RGB8);
+
+        // Trim each leg out of image
+        Rect left_rect(bbs[0].xmin, bbs[0].ymin,bbs[0].xmax-bbs[0].xmin,bbs[0].ymax-bbs[0].ymin);
+        Rect right_rect(bbs[1].xmin, bbs[1].ymin,bbs[1].xmax-bbs[1].xmin,bbs[1].ymax-bbs[1].ymin);
+        Mat left_pole = cv_ptr->image(left_rect);
+        Mat right_pole = cv_ptr->image(right_rect);
+
+        // Find gate's corner points
+        vector<Point2f> img_points;
+        if(!getPoints(left_pole, img_points)) {return false;}
+        if(!getPoints(right_pole, img_points)) {return false;}
+
+        // Transform Points back to main image
+        img_points[0].x += bbs[0].xmin;
+        img_points[0].y += bbs[0].ymin;
+        img_points[1].x += bbs[0].xmin;
+        img_points[1].y += bbs[0].ymin;
+        img_points[2].x += bbs[1].xmin;
+        img_points[2].y += bbs[1].ymin;
+        img_points[3].x += bbs[1].xmin;
+        img_points[3].y += bbs[1].ymin;
+
+        Mat rvec(3,1,cv::DataType<double>::type);
+        Mat tvec(3,1,cv::DataType<double>::type);
+
+        solvePnP(gate_truth_pts, img_points, occam_camera_matrix, occam_dist_coefs, rvec, tvec);
+        // cout << tvec << endl;
+        getPoseFromVectors(rvec, tvec, pose);
+        // cout << pose << endl;
+        class_name = "start_gate";
         return true;
     }
-    StartGateWatershed::StartGateWatershed()
+}
+
+void draw_on_image(Mat& image)
+{   
+    // Column then row
+    // loop through cols and rows (.cols, .rows)
+    for(int i=0; i < image.rows; i += 50)
+    {
+        line(image, Point(0,i), Point(image.cols,i), Scalar(0,0,255));
+    }
+    for(int j=0; j < image.cols; j += 50)
     {
-        ROS_INFO("Initializing StartGate Watershed!");
+        line(image, Point(j,0), Point(j,image.rows), Scalar(0,0,255));
     }
-}
\ No newline at end of file
+}
+
+int main(int argc, char **argv)
+{
+    cout << "Starting Watershed Testing" <<endl;
+    pose_generator::StartGateWatershed sgw;
+
+    // string image_name = "/home/luke/ros/robosub_ws/src/cusub/cusub_perception/localizer/src/frame0000.jpg";
+    string image_name = "/home/luke/ros/robosub_ws/src/cusub/cusub_perception/localizer/src/T0_10-27-18-image0_0137.jpg";
```

Comment:
Maybe add this image to the commit and make relative path? Just so others can test it out

Reply:
Know we went over this at the meeting, if you still want to see this in action, you can pull the most recent start_gate_3_legs branch, build, and run
$ rosrun localizer sg_2019
and you'll  see the image and point finding at each stage in the localization algorithm

Reply:
Adding new features to the start_gate_3_legs branch, if you'd like to visualize (and in case I need to come back to this someday) the last commit with development features is ad64b04230919549d8fc6876775d912014e6140d